cpu/kinetis/periph/i2c.c

```diff
@@ -45,474 +51,515 @@
-             * initiated by reading the data register (dev->D). */
-            dev->C1 &= ~I2C_C1_MST_MASK;
+        if (!read_flag) {
+            /* Send lower 8 bits of address */
+            res = i2c_tx_addr(dev, lsb);
         }
-
-        data[n] = dev->D;
-        TRACE("i2c: rx: %02x\n", (unsigned int)data[n]);
-        n++;
+        return res;
     }
-
-    return n;
+    /* 7 bit addressing */
+    /* Send address + R/W flag */
+    uint8_t addr_rw = (addr << 1);
+    if (read_flag) {
+        addr_rw |= I2C_READ;
+    }
+    return i2c_tx_addr(dev, addr_rw);
 }
 
-static inline int _i2c_transmit(I2C_Type *dev, const uint8_t *data, int length)
+int i2c_read_bytes(i2c_t dev, uint16_t addr, void *data, size_t len, uint8_t flags)
 {
-    int n = 0;
-
-    while (length > 0) {
-        TRACE("i2c: tx: %02x\n", data[n]);
-        dev->D = data[n];
-
-        while (!(dev->S & I2C_S_IICIF_MASK));
-
-        dev->S = I2C_S_IICIF_MASK;
+    assert((data != NULL) || (len == 0));
+    I2C_Type *i2c = i2c_config[dev].i2c;
+    DEBUG("i2c: r a:%02x, l:%u, f:%02x\n", addr, (unsigned)len, (unsigned)flags);
+    if (len == 1) {
+        /* Send NACK after next byte */
+        bit_set8(&i2c->C1, I2C_C1_TXAK_SHIFT);
+    }
+    else {
+        bit_clear8(&i2c->C1, I2C_C1_TXAK_SHIFT);
+    }
 
-        if (dev->S & I2C_S_RXAK_MASK) {
-            return n;
+    if (!(flags & I2C_NOSTART)) {
+        /* Send start condition and address */
+        int res = i2c_start(dev, addr, 1, flags);
+        if (res < 0) {
+            return res;
         }
-
-        n++;
-        length--;
     }
-
-    return n;
-}
-
-static inline void _i2c_stop(I2C_Type *dev)
-{
-    /* put bus in idle state */
-    dev->C1 = I2C_C1_IICEN_MASK;
-
-    /* wait for bus idle */
-    while (dev->S & I2C_S_BUSY_MASK);
-}
-
-static inline void _i2c_reset(I2C_Type *dev)
-{
-    /* put bus in idle state */
-    dev->C1 = I2C_C1_IICEN_MASK;
-    /* reset status flags */
-    dev->S = I2C_S_ARBL_MASK | I2C_S_IICIF_MASK;
-}
-
-
-int i2c_read_byte(i2c_t dev, uint8_t address, void *data)
-{
-    return i2c_read_bytes(dev, address, data, 1);
+    if (len > 0) {
+        /* We need IRQs enabled for interrupt based transfers */
+        assert(!__get_PRIMASK());
+        assert(!irq_is_in());
+        bit_clear8(&i2c->C1, I2C_C1_TX_SHIFT);
+        /* Configure IRQ transfer */
+        i2c_state[dev].rx.datap = data;
+        i2c_state[dev].rx.bytes_left = len;
+        i2c_state[dev].retval = 0;
+        bit_set8(&i2c->C1, I2C_C1_IICIE_SHIFT);
+        /* Initiate master receive mode by reading the data register once when
+         * the C1[TX] bit is cleared and C1[MST] is set */
+        volatile uint8_t dummy;
+        dummy = i2c->D;
+        ++dummy;
+        /* Wait until the ISR signals back */
+        TRACE("i2c: read C1=%02x S=%02x\n", (unsigned)i2c->C1, (unsigned)i2c->S);
+        thread_flags_t tflg = thread_flags_wait_any(THREAD_FLAG_KINETIS_I2C | THREAD_FLAG_TIMEOUT);
```

Comment:
We are checking for the TIMEOUT thread flag, but there is not yet anything setting it in the driver. I left this thread flag check here to make it simple in the future to add a timeout, but I did not yet find a satisfying solution which does not add extra dependencies to the driver. This part of the code can still be used if a user wants to eliminate the possibility of a bus hang blocking the application, they can use xtimer to provide a timer which sets the timeout flag.

Reply:
I'm fine with this, switching thread while waiting here is an excellent idea and IMHO we should do it for every MCU/drivers :)